File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/com/ririnto/sinon/harness/core/RuleContext.java
package com.ririnto.sinon.harness.core;

import com.ririnto.sinon.harness.Finding;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.Collection;
import java.util.List;

/**
 * Shared runtime context passed to Maven harness rules.
 */
public interface RuleContext {
    /**
     * Returns the absolute project root.
     *
     * @return project root path
     */
    Path root();

    /**
     * Returns the parsed manifest facade.
     *
     * @return manifest facade
     */
    Manifest manifest();

    /**
     * Returns the stack identifier for this context.
     *
     * @return stack name, such as "java"
     */
    String stack();

    /**
     * Reads a safe file path.
     *
     * @param path file path
     * @return file text
     * @throws MojoExecutionException when reading fails
     */
    String readSafe(Path path) throws MojoExecutionException;

    /**
     * Walks a safe file or directory.
     *
     * @param base file or directory path
     * @return walk result
     * @throws MojoExecutionException when walking fails
     */
    WalkResult walkSafe(Path base) throws MojoExecutionException;

    /**
     * Returns whether a symlink is allowed by the root contract.
     *
     * @param path path to inspect
     * @return true when the symlink is allowed
     */
    boolean isAllowedRootContractSymlink(Path path);

    /**
     * Returns source file paths for a manifest category and this context's stack.
     *
     * @param category manifest category name
     * @return list of source file paths
     * @throws MojoExecutionException when collection fails
     */
    List<Path> stackSources(String category) throws MojoExecutionException;

    /**
     * Collects findings for source root entries that are unsafe for a manifest category.
     *
     * @param category manifest category name
     * @return list of findings
     * @throws MojoExecutionException when validation fails
     */
    List<Finding> stackSourceFindings(String category) throws MojoExecutionException;

    /**
     * Safe walk result with filesystem findings.
     *
     * @param files safe files discovered
     * @param findings filesystem findings discovered while walking
     */
    record WalkResult(Collection<Path> files, Collection<Finding> findings) {
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/com/ririnto/sinon/harness/core/RuleContext.kt
package com.ririnto.sinon.harness.core

import com.ririnto.sinon.harness.ast.HarnessAstResults
import java.nio.file.Path

/**
 * Validation context passed to all rule implementations.
 */
interface RuleContext {
    /**
     * The project root path.
     */
    val root: Path

    /**
     * The harness validation manifest.
     */
    val manifest: Manifest

    /**
     * The current stack identifier (e.g. "kotlin", "typescript").
     */
    val stack: String

    /**
     * Safely read file content.
     *
     * @param path The file path relative to root.
     * @return File content; empty string if file does not exist or symlink is not allowed.
     */
    fun readSafe(path: String): String

    /**
     * Safely walk a directory tree.
     *
     * @param base The base directory path.
     * @return A pair of (list of files, list of findings for violations).
     */
    fun walkSafe(base: Path): WalkResult

    /**
     * Check if a symlink at the repository root is allowed per contract rules.
     *
     * @param path The symlink path.
     * @return true if allowed, false otherwise.
     */
    fun isAllowedRootContractSymlink(path: Path): Boolean

    /**
     * Get source root paths for a specific category and the current stack.
     *
     * @param category The category name.
     * @return List of expanded Path objects; empty if not found or stack not in manifest.
     */
    fun stackSources(category: String): List<Path>

    /**
     * Get source-root safety findings for a specific category.
     *
     * @param category The category name.
     * @return Findings describing unsafe source roots (absolute roots, traversal roots, symlink roots, invalid globs).
     */
    fun stackSourceFindings(category: String): List<HarnessAstResults.Finding>

    /**
     * Result of a safe directory walk.
     */
    data class WalkResult(
        /**
         * Paths of files found during the walk.
         */
        val paths: List<Path>,
        /**
         * Findings for symlink violations or access errors.
         */
        val findings: List<HarnessAstResults.Finding>,
    )
}
